### PR TITLE
Add CHGNet model support

### PR DESCRIPTION
**Pull request by brunocucco**
This PR adds a wrapper for the CHGNet model (Deng et al., Nat. Mach. Intell 5, 1031–1041, 2023) to TorchSim. CHGNet is a pretrained universal neural network potential for charge-informed atomistic modeling.

## Changes Made
- **Added a new model**: `torch_sim/models/chgnet.py` - CHGNetModel wrapper
- **Added tests**: `tests/models/test_chgnet.py` - Comprehensive test suite
- **Added example**: `examples/scripts/1_Introduction/1.4_CHGNet.py` - Demonstration
- **Added dependencies**: Added `chgnet>=0.4.2` to `pyproject.toml`

## Features
- Inherits from ModelInterface (consistent with other models)
- Supports batched calculations for multiple systems
- Handles periodic boundary conditions correctly
- Integrates with TorchSim workflows (optimization, dynamics)

## Testing
- All tests pass: `python -m pytest tests/models/test_chgnet.py -v`
- Example script runs successfully and results are in-line with MACE.

## Notes
- CHGNet uses different energy scales than other models (Handled and documented at `torch_sim/models/chgnet.py`)
- Implementation follows TorchSim patterns already in-place for MACE and FairChem.
- Example script demonstrates CHGNet vs MACE comparison showing both models correctly minimize forces and predict similar values.

## How to Test
```bash
pip install chgnet>=0.4.2
python -m pytest tests/models/test_chgnet.py -v
python examples/scripts/1_Introduction/1.4_CHGNet.py
```

### CODE REVIEW

File: examples/scripts/1_Introduction/1.4_CHGNet.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
+"""CHGNet model example for TorchSim."""
+
+# /// script
+# dependencies = ["chgnet>=0.4.2"]
```
Comment (abhijeetgangan) [inline examples/scripts/1_Introduction/1.4_CHGNet.py]: Since this script would compare with mace it would make sense to add mace here and also remove the try except so that it can run in CI?

File: examples/scripts/1_Introduction/1.4_CHGNet.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+            f"{mace_final_force:<15.6f}"
+        )
+    else:
+        print(
+            f"{system_name:<10} {chgnet_initial_energy:<15.6f} "
+            f"{chgnet_final_energy:<15.6f} {chgnet_final_force:<15.6f} "
+            f"{'N/A':<15} {'N/A':<15} {'N/A':<15}"
+        )
+
+print("\n" + "=" * 100)
```
Comment (abhijeetgangan) [inline examples/scripts/1_Introduction/1.4_CHGNet.py]: Could do with less printing.

File: tests/models/test_chgnet.py [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+    )
+
+
+@pytest.mark.parametrize("dtype", [torch.float32, torch.float64])
+def test_chgnet_dtype_working(si_atoms: Atoms, dtype: torch.dtype) -> None:
+    """Test that CHGNet works with different dtypes."""
+    model = CHGNetModel(
+        device=DEVICE,
+        dtype=dtype,
+        compute_forces=True,
```
Comment (abhijeetgangan) [inline tests/models/test_chgnet.py]: Shoudn't the compute stress be true here since you are checking the dtype for the output stress?

File: tests/models/test_chgnet.py [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+    assert result["forces"].dtype == dtype
+    assert result["stress"].dtype == dtype
+
+
+def test_chgnet_batched_calculations() -> None:
+    """Test that CHGNet handles batched calculations correctly."""
+    from ase.build import bulk
+
+    # Create multiple systems
+    si_atoms = bulk("Si", "diamond", a=5.43, cubic=True)
```
Comment (abhijeetgangan) [inline tests/models/test_chgnet.py]: Could you use the existing fixtures to define the atoms?

File: torch_sim/models/chgnet.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+                torch.tensor(total_energy, device=self.device, dtype=self.dtype)
+            )
+
+        results["energy"] = torch.stack(energies)
+
+        # Process forces
+        if self.compute_forces:
+            forces_list = [
+                torch.tensor(result["f"], device=self.device, dtype=self.dtype)
+                for result in chgnet_results
```
Comment (abhijeetgangan) [inline torch_sim/models/chgnet.py:176]: CHGNet doesn't output a batched tensor?

**Comment by abhijeetgangan** [comment]
Thanks for the contribution! I left some comments and happy to merge once those are added. Also, it would be good to the model in CI in `test.yml` in `test-model`. On a side note would it make sense to add other outputs of CHGNet?

File: tests/models/test_chgnet.py [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+
+    # Check that results have correct dtype
+    assert result["energy"].dtype == dtype
+    assert result["forces"].dtype == dtype
+    assert result["stress"].dtype == dtype
+
+
+def test_chgnet_batched_calculations() -> None:
+    """Test that CHGNet handles batched calculations correctly."""
+    from ase.build import bulk
```
Comment (orionarcher) [inline tests/models/test_chgnet.py]: Many of the tests here are redundant with the tests in `torch_sim.models.tests.make_validate_model_outputs_test` and `torch_sim.models.tests.make_model_calculator_consistency_test`. I'd recommend taking a look at how some of the other models are tested ([like mattersim](https://github.com/TorchSim/torch-sim/blob/main/tests/models/test_mattersim.py)) and following that paradigm.

**Comment by brunocucco** [comment]
Hi, 

Thank you for all the comments and suggestions, happy to contribute. I've been working on those and will update the PR as soon as possible. Best.

File: torch_sim/models/chgnet.py (hunk 5 above)
Comment (brunocucco) [reply torch_sim/models/chgnet.py:176]: Unfortunately, CHGNet's doesn't return batched tensors directly. It returns a list of dictionaries, so I had to manually convert and concatenate them. The current implementation does this:
- Converts each result to a tensor
- Concatenate with torch.cat() for forces
- Stack with torch.stack() for stress

File: tests/models/test_chgnet.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    from chgnet.model.model import CHGNet
+
+    from torch_sim.models.chgnet import CHGNetModel
+except (ImportError, ValueError):
+    pytest.skip(
+        f"CHGNet not installed: {traceback.format_exc()}", allow_module_level=True
+    )
+
+
+class CHGNetCalculator(Calculator):
```
Comment (abhijeetgangan) [inline tests/models/test_chgnet.py]: Looks like CHGNet has an ase calculator - https://github.com/CederGroupHub/chgnet/blob/ef175c424da35194872403c11b18665d4fed5de8/chgnet/model/dynamics.py#L58

**Comment by abhijeetgangan** [comment]
Looks like the test for chgnet are failing. Good to merge once the comment is resolved and the failing tests are fixed.

File: torch_sim/models/chgnet.py [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+
+        sim_state = (
+            state
+            if isinstance(state, ts.SimState)
+            else ts.SimState(**state, masses=torch.ones_like(state["positions"]))
+        )
+
+        # Validate that atomic numbers
+        if sim_state.atomic_numbers is None:
+            raise ValueError("Atomic numbers must be provided in the state for CHGNet.")
```
Comment (orionarcher) [inline torch_sim/models/chgnet.py]: Atomic numbers are generally required by TorchSim so I don't think we need to specifically assert their existence here. Can also delete the related test.

File: tests/models/test_chgnet.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+    state_dict = {
+        "positions": torch.randn(8, 3, device=DEVICE, dtype=DTYPE),
+        "cell": torch.eye(3, device=DEVICE, dtype=DTYPE).unsqueeze(0),
+        "pbc": True,
+        "atomic_numbers": None,  # Missing atomic numbers
+        "system_idx": torch.zeros(8, dtype=torch.long, device=DEVICE),
+    }
+
+    with pytest.raises(ValueError, match="Atomic numbers must be provided"):
+        model.forward(state_dict)
```
Comment (orionarcher) [inline tests/models/test_chgnet.py]: Dont need this test, atomic numbers should never be none

**Comment by abhijeetgangan** [comment]
Looks like the CHGnet is moved to [matgl](https://github.com/CederGroupHub/chgnet/commit/ef175c424da35194872403c11b18665d4fed5de8) as per the latest changes on that repo. Maybe we should adapt to that repo?

Also, the tests are failing on mac os.

**Comment by brunocucco** [comment]
As an update, I will take a look on that as soon as possible. I've been a bit busy in the past few weeks. Best.

**Comment by CompRhys** [comment]
Closing based on #120 and desired move towards External model posture.
